ext/mruby_engine/dlmalloc.c

```diff
@@ -4523,6 +4527,40 @@ struct mallinfo mspace_mallinfo(mspace msp) {
   }
   return internal_mallinfo(ms);
 }
+
+void log_stuff(mspace msp) {
+  mstate m = (mstate)msp;
+  ensure_initialization();
+  if (!PREACTION(m)) {
+    check_malloc_state(m);
+    if (is_initialized(m)) {
+      size_t nfree = SIZE_T_ONE; /* top always free */
+      size_t mfree = m->topsize + TOP_FOOT_SIZE;
+      size_t sum = mfree;
+      msegmentptr s = &m->seg;
+      while (s != 0) {
+        fprintf(stderr, "S @ %p\n", (void *)s);
+        mchunkptr q = align_as_chunk(s->base);
+        while (segment_holds(s, q) &&
+               q != m->top && q->head != FENCEPOST_HEAD) {
+          size_t sz = chunksize(q);
+          sum += sz;
+          if (!is_inuse(q)) {
+            mfree += sz;
+            ++nfree;
+          }
+          q = next_chunk(q);
+        }
+        s = s->next;
+      }
+
+      fprintf(stderr, "%zu, %zu, %zu, %zu, %zu, %zu, %zu\n",
+          sum, nfree, m->footprint - sum, m->max_footprint, m->footprint - mfree, mfree, m->topsize);
```

Comment:
Right, but overflow _is_ just maths... now, all good programmers are lazy, I'll give you that! :D — as an aside there are some warnings wrt narrowing because of the all the madness with the macros (@sgnr hinted to those in the past). Now I quickly reviewed them and _I think_ they are all harmless (we shouldn't ever request more than `(unsigned int) max_value`... Or we'd be in trouble anyways (regardless of the narrowing). But anyways, if anyone wants to double check these as well, that'd be great.